serializer_impl.hh

```diff
@@ -877,30 +923,52 @@ public:
         }
     };
 
+
     using const_iterator = iterator;
 
     static_assert(std::input_iterator<iterator>);
     static_assert(std::sentinel_for<iterator, iterator>);
 
-    iterator begin() noexcept {
-        return iterator(_in, 0);
+    iterator begin() noexcept requires(IsForward) {
+        return {_in, 0};
+    }
+    const_iterator begin() const noexcept requires(IsForward) {
+        return {_in, 0};
+    }
+    const_iterator cbegin() const noexcept requires(IsForward) {
+        return {_in, 0};
+    }
+
+    iterator end() noexcept requires(IsForward) {
+        return {_in, _size};
     }
-    const_iterator begin() const noexcept {
-        return const_iterator(_in, 0);
+    const_iterator end() const noexcept requires(IsForward) {
+        return {_in, _size};
     }
-    const_iterator cbegin() const noexcept {
-        return const_iterator(_in, 0);
+    const_iterator cend() const noexcept requires(IsForward) {
+        return {_in, _size};
     }
 
-    iterator end() noexcept {
-        return iterator(_in, _size);
+    iterator begin() noexcept requires(!IsForward) {
+        return {_substreams.crbegin(), 0};
```

Comment:
It's worth adding a comment about the idx logic, maybe arounf the reverse iterator data class.
Since it seems to be the idx from _substreams rbegin().

Reply:
idx is used the same as it was in forward iterator. I can add a comment though anyway.